// @ldo/solid-react/dist/document/FetchableDocument.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FetchableDocument = void 0;
const events_1 = __importDefault(require("events"));
const STATE_UPDATE = "stateUpdate";
class FetchableDocument extends events_1.default {
    constructor(dependencies) {
        super();
        this._isLoading = false;
        this._isWriting = false;
        this._didInitialFetch = false;
        this.dependencies = dependencies;
    }
    get isLoading() {
        return this._isLoading;
    }
    get didInitialFetch() {
        return this._didInitialFetch;
    }
    get error() {
        return this._error;
    }
    get isWriting() {
        return this._isWriting;
    }
    get onDocumentError() {
        return this.dependencies.onDocumentError;
    }
    async read() {
        this._isLoading = true;
        this.emitStateUpdate();
        const documentError = await this.fetchDocument();
        this._isLoading = false;
        this._didInitialFetch = true;
        if (documentError) {
            this.setError(documentError);
        }
        this.emitStateUpdate();
    }
    async reload() {
        return this.read();
    }
    beginWrite() {
        this._isWriting = true;
        this.emitStateUpdate();
    }
    endWrite(error) {
        if (error) {
            this.setError(error);
        }
        this._isWriting = false;
        this.emitStateUpdate();
    }
    setError(error) {
        this._error = error;
        this.emitStateUpdate();
        if (this.onDocumentError) {
            this.onDocumentError(error);
        }
    }
    emitStateUpdate() {
        this.emit(STATE_UPDATE);
    }
    onStateUpdate(callback) {
        this.on(STATE_UPDATE, callback);
    }
    offStateUpdate(callback) {
        this.off(STATE_UPDATE, callback);
    }
}
exports.FetchableDocument = FetchableDocument;
