// @ldo/subscribable-dataset/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WrapperSubscribableDatasetFactory = exports.WrapperSubscribableDataset = exports.ProxyTransactionalDataset = exports.serializedToSubscribableDataset = exports.createSubscribableDatasetFactory = exports.createSubscribableDataset = void 0;
var createWrapperSubscribableDataset_1 = require("./createWrapperSubscribableDataset");
Object.defineProperty(exports, "createSubscribableDataset", { enumerable: true, get: function () { return __importDefault(createWrapperSubscribableDataset_1).default; } });
Object.defineProperty(exports, "createSubscribableDatasetFactory", { enumerable: true, get: function () { return createWrapperSubscribableDataset_1.createWrapperSubscribableDatasetFactory; } });
var createWrapperSubscribableDatasetFromSerializedInput_1 = require("./createWrapperSubscribableDatasetFromSerializedInput");
Object.defineProperty(exports, "serializedToSubscribableDataset", { enumerable: true, get: function () { return __importDefault(createWrapperSubscribableDatasetFromSerializedInput_1).default; } });
var ProxyTransactionalDataset_1 = require("./ProxyTransactionalDataset");
Object.defineProperty(exports, "ProxyTransactionalDataset", { enumerable: true, get: function () { return __importDefault(ProxyTransactionalDataset_1).default; } });
var WrapperSubscribableDataset_1 = require("./WrapperSubscribableDataset");
Object.defineProperty(exports, "WrapperSubscribableDataset", { enumerable: true, get: function () { return __importDefault(WrapperSubscribableDataset_1).default; } });
var WrapperSubscribableDatasetFactory_1 = require("./WrapperSubscribableDatasetFactory");
Object.defineProperty(exports, "WrapperSubscribableDatasetFactory", { enumerable: true, get: function () { return __importDefault(WrapperSubscribableDatasetFactory_1).default; } });
__exportStar(require("./types"), exports);
__exportStar(require("./mergeDatasetChanges"), exports);
