// @ldo/solid-react/dist/document/resource/Resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Resource = void 0;
const FetchableDocument_1 = require("../FetchableDocument");
const DocumentFetchError_1 = require("../errors/DocumentFetchError");
class Resource extends FetchableDocument_1.FetchableDocument {
    constructor(uri, dependencies) {
        super(dependencies);
        this.uri = uri;
        this.dependencies1 = dependencies;
    }
    get accessRules() {
        return this.accessRulesStore.get(this);
    }
    get parentContainer() {
        return this.containerResourceStore.getContainerForResouce(this);
    }
    get ["@id"]() {
        return this.uri;
    }
    get fetch() {
        return this.dependencies1.fetch;
    }
    get accessRulesStore() {
        return this.dependencies1.accessRulesStore;
    }
    get containerResourceStore() {
        return this.dependencies1.containerResourceStore;
    }
    async delete() {
        this.beginWrite();
        const response = await this.fetch(this.uri, {
            method: "DELETE",
        });
        if (response.status >= 200 && response.status < 300) {
            this.endWrite();
            this.parentContainer?.removeContainedResources(this);
            return;
        }
        this.endWrite(new DocumentFetchError_1.DocumentFetchError(this, response.status, `Could not delete ${this.uri}`));
    }
    static normalizeUri(uri) {
        const [strippedHashUri] = uri.split("#");
        return strippedHashUri;
    }
}
exports.Resource = Resource;
