// @ldo/jsonld-dataset-proxy/dist/ContextUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextUtil = void 0;
const shorthandToIriMap = {
    "@type": "http://www.w3.org/1999/02/22-rdf-syntax-ns#type",
};
class ContextUtil {
    constructor(context) {
        this.context = context;
        this.iriToKeyMap = {};
        Object.entries(context).forEach(([contextKey, contextValue]) => {
            if (typeof contextValue === "string") {
                this.iriToKeyMap[this.keyIdToIri(contextValue)] = contextKey;
            }
            else if (typeof contextValue === "object" &&
                contextValue["@id"]) {
                this.iriToKeyMap[this.keyIdToIri(contextValue["@id"])] =
                    contextKey;
            }
        });
    }
    keyToIri(key) {
        if (!this.context[key]) {
            return key;
        }
        else if (typeof this.context[key] === "string") {
            return this.keyIdToIri(this.context[key]);
        }
        else if (this.context[key] && this.context[key]["@id"]) {
            return this.keyIdToIri(this.context[key]["@id"]);
        }
        return key;
    }
    keyIdToIri(keyId) {
        if (shorthandToIriMap[keyId]) {
            return shorthandToIriMap[keyId];
        }
        else {
            return keyId;
        }
    }
    iriToKey(iri) {
        if (this.iriToKeyMap[iri]) {
            return this.iriToKeyMap[iri];
        }
        return iri;
    }
    getType(key) {
        if (typeof this.context[key] === "object" &&
            this.context[key]["@type"]) {
            return this.context[key]["@type"];
        }
        return "http://www.w3.org/2001/XMLSchema#string";
    }
    isArray(key) {
        return !!(this.context[key] &&
            typeof this.context[key] === "object" &&
            this.context[key]["@container"] &&
            this.context[key]["@container"] === "@set");
    }
    isLangString(key) {
        return !!(this.context[key] &&
            typeof this.context[key] === "object" &&
            this.context[key]["@type"] &&
            this.context[key]["@type"] ===
                "http://www.w3.org/1999/02/22-rdf-syntax-ns#langString");
    }
}
exports.ContextUtil = ContextUtil;
