// @ldo/solid-react/dist/useLdoMethods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUseLdoMethods = void 0;
const ldo_1 = require("@ldo/ldo");
const ldo_2 = require("@ldo/ldo");
const ldo_3 = require("@ldo/ldo");
function createUseLdoMethods(dataset) {
    return {
        dataset: dataset,
        getResource: dataset.getResource.bind(dataset),
        getSubject(shapeType, subject) {
            return dataset.usingType(shapeType).fromSubject(subject);
        },
        createData(shapeType, subject, ...resources) {
            const linkedDataObject = dataset
                .usingType(shapeType)
                .write(...resources.map((r) => r.uri))
                .fromSubject(subject);
            (0, ldo_3.startTransaction)(linkedDataObject);
            return linkedDataObject;
        },
        changeData(input, ...resources) {
            const [transactionLdo] = (0, ldo_2.write)(...resources.map((r) => r.uri)).usingCopy(input);
            (0, ldo_3.startTransaction)(transactionLdo);
            return transactionLdo;
        },
        commitData(input) {
            const changes = (0, ldo_1.transactionChanges)(input);
            return dataset.commitChangesToPod(changes);
        },
    };
}
exports.createUseLdoMethods = createUseLdoMethods;
