// @ldo/solid-react/dist/document/accessRules/AccessRules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessRules = void 0;
const solid_client_1 = require("@inrupt/solid-client");
const FetchableDocument_1 = require("../FetchableDocument");
const DocumentError_1 = require("../errors/DocumentError");
class AccessRules extends FetchableDocument_1.FetchableDocument {
    constructor(resource, dependencies) {
        super(dependencies);
        this._publicAccess = null;
        this._agentAccess = null;
        this.dependencies0 = dependencies;
        this.resource = resource;
    }
    get publicAccess() {
        return this._publicAccess;
    }
    get agentAccess() {
        return this._agentAccess;
    }
    get fetch() {
        return this.dependencies0.fetch;
    }
    async fetchDocument() {
        try {
            const [publicAccess, agentAccess] = await Promise.all([
                solid_client_1.universalAccess.getPublicAccess(this.resource.uri, {
                    fetch: this.fetch,
                }),
                solid_client_1.universalAccess.getAgentAccessAll(this.resource.uri, {
                    fetch: this.fetch,
                }),
            ]);
            this._publicAccess = publicAccess || {
                read: false,
                write: false,
                append: false,
                controlRead: false,
                controlWrite: false,
            };
            this._agentAccess = agentAccess || {};
            return undefined;
        }
        catch (err) {
            if (typeof err === "object" && err.message) {
                this.setError(new DocumentError_1.DocumentError(this, err.message));
            }
            this.setError(new DocumentError_1.DocumentError(this, "Error Fetching Access Rules"));
        }
    }
}
exports.AccessRules = AccessRules;
