// @ldo/solid/dist/requester/ContainerRequester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerRequester = exports.IS_ROOT_CONTAINER_KEY = void 0;
const Requester_1 = require("./Requester");
const checkRootContainer_1 = require("./requests/checkRootContainer");
exports.IS_ROOT_CONTAINER_KEY = "isRootContainer";
class ContainerRequester extends Requester_1.Requester {
    constructor(uri, context) {
        super(context);
        this.uri = uri;
    }
    read() {
        return super.read();
    }
    createDataResource(overwrite) {
        return super.createDataResource(overwrite);
    }
    async isRootContainer() {
        return this.requestBatcher.queueProcess({
            name: exports.IS_ROOT_CONTAINER_KEY,
            args: [this.uri, { fetch: this.context.fetch }],
            perform: checkRootContainer_1.checkRootContainer,
            modifyQueue: (queue, currentlyLoading) => {
                if (queue.length === 0 &&
                    currentlyLoading?.name === exports.IS_ROOT_CONTAINER_KEY) {
                    return currentlyLoading;
                }
                else if (queue[queue.length - 1]?.name === exports.IS_ROOT_CONTAINER_KEY) {
                    return queue[queue.length - 1];
                }
                return undefined;
            },
        });
    }
}
exports.ContainerRequester = ContainerRequester;
